File: srcs/juloo/keyboard2/CustomView.java
package juloo.keyboard2;

import android.content.Context;
import android.view.View;
import android.view.MotionEvent;
import android.graphics.Canvas;
import juloo.javacaml.Caml;
import juloo.javacaml.Value;

public class CustomView extends View
{
	Value				callbacks;

	public CustomView(Context context) { super(context); }

	static final int m_id_onTouchEvent = Caml.hashVariant("onTouchEvent");
	@Override
	public boolean		onTouchEvent(MotionEvent event)
	{
		Caml.method(callbacks, m_id_onTouchEvent);
		Caml.argObject(event);
		return Caml.callBool();
	}

	static final int m_id_onMeasure = Caml.hashVariant("onMeasure");
	@Override
	public void			onMeasure(int wSpec, int hSpec)
	{
		Caml.method(callbacks, m_id_onMeasure);
		Caml.argInt(wSpec);
		Caml.argInt(hSpec);
		Caml.callUnit();
	}

	static final int m_id_onDraw = Caml.hashVariant("onDraw");
	@Override
	protected void		onDraw(Canvas canvas)
	{
		Caml.method(callbacks, m_id_onDraw);
		Caml.argObject(canvas);
		Caml.callUnit();
	}

	static final int m_id_onDetachedFromWindow = Caml.hashVariant("onDetachedFromWindow");
	@Override
	public void			onDetachedFromWindow()
	{
		Caml.method(callbacks, m_id_onDetachedFromWindow);
		Caml.callUnit();
	}
}

File: srcs/android_utils/CustomView.ml
open Android_api.Content
open Android_api.Graphics
open Android_api.View

type callbacks = <
	onTouchEvent : Motion_event.t -> bool;
	onMeasure : int -> int -> unit;
	onDraw : Canvas.t -> unit;
	onDetachedFromWindow : unit
>

let create (context : _ Context.t')
		(f : View.t -> callbacks) : View.t =
	let cls = Jclass.find_class "juloo/keyboard2/CustomView" in
	let obj_field = Jclass.get_field cls "callbacks" "Ljuloo/javacaml/Value;" in
	let constr = Jclass.get_constructor cls "(Landroid/content/Context;)V" in
	Jcall.push_object context;
	let obj = Jcall.new_ cls constr in
	let callbacks = f obj in
	Jcall.write_field_value obj obj_field callbacks;
	obj
